In [3]:
import sys
print(sys.version)

3.10.0 (v3.10.0:b494f5935c, Oct  4 2021, 14:59:20) [Clang 12.0.5 (clang-1205.0.22.11)]


# Python 3.9 추가사항

- Dictionary 순서 보존 
- Dictionary 병합 연산자 |

In [7]:
google = {'android': 11}
apple = {'ios': 14}

print(google | apple)
print(apple | google)

{'android': 11, 'ios': 14}
{'ios': 14, 'android': 11}


- 접두사 삭제 함수 removeprefix()
- 접미사 삭제 함수 removesuffix()

In [6]:
string = 'eicheich'
string = string.removeprefix('eich')

print(string)

eicheich


In [15]:
string = 'eichmann'
string = string.removesuffix('mann')

print(string)

eich


In [13]:
import random

a = [1,2,3]

random.sample(a, k=len(a))

[1, 3, 2]

#### 기타 변경사항

- math.gcd(), math.lcm() 파라미터 2개 이상 받음
- import 에러 ValueError 대신 ImportError 발생
- random.shuffle() 함수 삭제예정 (3.11) -> random.sample() 사용 권장
- 일부 내장 함수 최적화

# Python 3.10 추가사항

- 타입 힌트 개선

In [34]:
from typing import Union

def square(number: Union[int, float]) -> Union[int, float]:
    return number ** 2

print(square(2.1))

4.41


In [35]:
def square(number: int | float) -> int | float:
    return number ** 2

print(square(2.1))

4.41


In [26]:
from typing import List

def add(numbers: List[int]) -> int:
    return sum(numbers)

In [39]:
print(add([1,2,3]))

6


In [42]:
def add(numbers: list[int]) -> int:
    return sum(numbers)

In [43]:
print(add([1,2,3]))

6


- 에러 메시지 개선

In [45]:
expected = {1,2,3,4

SyntaxError: '{' was never closed (3739790726.py, line 1)

In [48]:
if 1 = True:
    print('hello')

SyntaxError: cannot assign to literal here. Maybe you meant '==' instead of '='? (3969185522.py, line 1)

In [49]:
if 1 == True:
print('hello')

IndentationError: expected an indented block after 'if' statement on line 1 (3708491032.py, line 2)

- 패턴 매칭 기능 추가
- JAVA swith case 문과 유사

In [18]:
string = 'abcd'

match string:
    case 'abcd':
        print(1)
    case 'abc' | 'abd':
        print(2)
    case 'abcd':
        print(3)

1


In [21]:
string = 'abcd'

match string:
    case 'ab':
        print(1)
    case 'abc' | 'abd':
        print(2)
    case _:
        print('wile card')

wile card


## Effective Python

- 정렬된 리스트 탐색시 bisect 사용 권장

In [29]:
import bisect

arr = list(range(0,500,3))

In [30]:
arr

[0,
 3,
 6,
 9,
 12,
 15,
 18,
 21,
 24,
 27,
 30,
 33,
 36,
 39,
 42,
 45,
 48,
 51,
 54,
 57,
 60,
 63,
 66,
 69,
 72,
 75,
 78,
 81,
 84,
 87,
 90,
 93,
 96,
 99,
 102,
 105,
 108,
 111,
 114,
 117,
 120,
 123,
 126,
 129,
 132,
 135,
 138,
 141,
 144,
 147,
 150,
 153,
 156,
 159,
 162,
 165,
 168,
 171,
 174,
 177,
 180,
 183,
 186,
 189,
 192,
 195,
 198,
 201,
 204,
 207,
 210,
 213,
 216,
 219,
 222,
 225,
 228,
 231,
 234,
 237,
 240,
 243,
 246,
 249,
 252,
 255,
 258,
 261,
 264,
 267,
 270,
 273,
 276,
 279,
 282,
 285,
 288,
 291,
 294,
 297,
 300,
 303,
 306,
 309,
 312,
 315,
 318,
 321,
 324,
 327,
 330,
 333,
 336,
 339,
 342,
 345,
 348,
 351,
 354,
 357,
 360,
 363,
 366,
 369,
 372,
 375,
 378,
 381,
 384,
 387,
 390,
 393,
 396,
 399,
 402,
 405,
 408,
 411,
 414,
 417,
 420,
 423,
 426,
 429,
 432,
 435,
 438,
 441,
 444,
 447,
 450,
 453,
 456,
 459,
 462,
 465,
 468,
 471,
 474,
 477,
 480,
 483,
 486,
 489,
 492,
 495,
 498]

In [78]:
print(bisect.bisect_left(arr, 10))

# 존재하지 않는 경우 순서상 해당 값을 삽입해야 할 자리의 인덱스 반환
print(bisect.bisect_left(arr, 5))

4
2


In [90]:
%%timeit -n 10000 -r 1000

arr.index(201)

751 ns ± 95.3 ns per loop (mean ± std. dev. of 1000 runs, 10000 loops each)


In [91]:
%%timeit -n 10000 -r 1000

bisect.bisect_left(arr, 201)

# list.index() 를 사용할 이유가 없다.
# 코딩테스트때 유용할 것 같다.

190 ns ± 32.3 ns per loop (mean ± std. dev. of 1000 runs, 10000 loops each)


- 사후 디버깅
- 에러가 난 위치를 모를경우
- pdb.pm()

In [31]:
import pdb

a = 1
b = '1'
pdb.set_trace()
print(a+b)

--Return--
None
> /var/folders/cc/nx0n5xgd0p12q50x_6qkqgy80000gn/T/ipykernel_4764/1504711619.py(5)<module>()
      2 
      3 a = 1
      4 b = '1'
----> 5 pdb.set_trace()
      6 print(a+b)

ipdb> a
ipdb> print(a)
1
ipdb> print(b)
1
ipdb> a+b
ipdb> print(a+b)
*** TypeError: unsupported operand type(s) for +: 'int' and 'str'
ipdb> quit()


BdbQuit: 

In [33]:
import pdb

pdb.pm()

a = 1
b = '1'

print(a+b)

## 또는 터미널에서 실행시 `python3 -m pdb -c continue FILE_NAME.py` 사용

None
> /var/folders/cc/nx0n5xgd0p12q50x_6qkqgy80000gn/T/ipykernel_4764/524956487.py(8)<module>()
      6 b = '1'
      7 
----> 8 print(a+b)
      9 
     10 ## 또는 터미널에서 실행시 `python3 -m pdb -c continue FILE_NAME.py` 사용

ipdb> print(a)
1
ipdb> print(b)
1
ipdb> quit()


TypeError: unsupported operand type(s) for +: 'int' and 'str'

- 언패킹 기법

In [107]:
arr = [1,2,3,4,5]

print(arr[0], arr[-1])

a,b,c,d,e = arr

print(a,e)

1 5
1 5


In [108]:
a, *b, c = arr

print(a, b, c)

1 [2, 3, 4] 5


In [109]:
a, b, *c = arr

print(a, b, c)

1 2 [3, 4, 5]


- 병렬성을 위해 subporcess 사용

In [111]:
import os

os.system('echo 123')

123


0

In [35]:
import subprocess

result = subprocess.run(['echo', 'hi'],
                       capture_output=True,
                       encoding='utf-8')

result = subprocess.run(['echo', 'hi'],
                       capture_output=True,
                       encoding='utf-8')

result = subprocess.run(['echo', 'hi'],
                       capture_output=True,
                       encoding='utf-8')

# 파이선 인터프리터와 독립적으로 실행됨

In [40]:
import multiprocessing as mp

In [151]:
import time

proc = subprocess.Popen(['sleep', '5'])

while proc.poll() is None:
    time.sleep(1)
    print('작업중')
    
print('작업완료')

작업중
작업중
작업중
작업중
작업중
작업완료


- 제너레이터 사용

In [44]:
arr = [1,2,3,...]

In [45]:
list(map(lambda x: x*2, arr))

[2, 4, 6]

In [46]:
[v*2 for v in arr]

[2, 4, 6]

In [47]:
(v*2 for v in arr)

2
4
6


In [49]:
def gen():
    for v in arr:
        yield v*2

In [50]:
for a in gen():
    print(a)

2
4
6


In [11]:
class Gen:
    def __init__(self):
        self.arr = [1,2,3,4]
    
    def __iter__(self):
        for v in self.arr:
            yield v

In [12]:
a = Gen()

In [14]:
for v in a:
    print(v)

1
2
3
4


In [51]:
import sys

In [53]:
arr = list(range(1,100000000))

In [55]:
gen = (i for i in range(1,100000000))

In [54]:
sys.getsizeof(arr)

800000048

In [57]:
sys.getsizeof(gen)

104

In [ ]:
50

5 5 5 5 5 5 5 5 5 5 5